python/katana_setup.py

```diff
@@ -0,0 +1,220 @@
+    return files
+
+
+def module_name_from_source_name(f, source_root_name):
+    parts = []
+    for p in f.parents:
+        if str(p.name) == source_root_name:
+            break
+        parts.insert(0, p.name)
+    parts.append(f.stem)
+    module_name = ".".join(parts)
+    return module_name
+
+
+def process_jinja_file(filename, build_source_root):
+    output_file = build_source_root / filename.parent / filename.stem
+    output_file.parent.mkdir(parents=True, exist_ok=True)
+    for p in filename.parents:
+        init_file = build_source_root / p / "__init__.pxd"
+        init_file.touch(exist_ok=True)
+
+    # TODO(amp): Ideally this would only process the jinja file when the inputs change. But it's fast and I'm lazy.
+    #  https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.meta.find_referenced_templates
+    regenerated = generate_from_jinja.run("python" / filename.parent, filename.name, output_file)
+    if regenerated:
+        print(f"Processed {filename} with Jinja2.")
+    return output_file
+
+
+def collect_cython_files(source_root):
+    """
+    Search `source_root` for pyx and pxd files and jinja files which generate them
+    and build lists of the existing and generated pyx and pxd files.
+
+    :param source_root: The python source root to search.
+    :return: pxd_files, pyx_files
+    """
+    source_root = Path(source_root)
+    pyx_jinja_files = find_files(source_root, source_root.parent, ".pyx.jinja")
+    pxd_jinja_files = find_files(source_root, source_root.parent, ".pxd.jinja")
+
+    pxd_files = find_files(source_root, "", ".pxd")
+    pyx_files = find_files(source_root, "", ".pyx")
+
+    for f in pyx_jinja_files + pxd_jinja_files:
+        output_file: Path = process_jinja_file(f, source_root.parent)
+        if output_file.suffix == ".pyx":
+            pyx_files.append(output_file)
+        else:
+            pxd_files.append(output_file)
+
+    assert all("pyx" in f.name for f in pyx_files)
+    assert all("pxd" in f.name for f in pxd_files)
+
+    return pxd_files, pyx_files
+
+
+def _build_cython_extensions(pyx_files, source_root_name, extension_options):
+    cython_extensions = []
+    for f in pyx_files:
+        module_name = module_name_from_source_name(f, source_root_name)
+        cython_extensions.append(setuptools.Extension(module_name, [str(f)], language="c++", **extension_options))
+
+    return cython_extensions
+
+
+def cythonize(module_list, *, source_root, **kwargs):
+    extension_options = load_lang_config("CXX")
+    extension_options["include_dirs"].append(numpy.get_include())
+
+    if not extension_options["extra_compile_args"]:
+        extension_options["extra_compile_args"] = ["-std=c++17", "-Werror"]
+
+    if extension_options["compiler"]:
+        compiler = " ".join(extension_options["compiler"])
+        os.environ["CXX"] = compiler
+        os.environ["CC"] = compiler
+
+        # Because of odd handling of the linker in setuptools with C++ the
```

Comment:
It is probably for the best we use the same program. We can get into issues mixing programs from different toolchains: https://gist.github.com/jlblancoc/99521194aba975286c80f93e47966dc5

Reply:
I'm not sure what you are referring to here. Am I already doing what you are asking for? Or do I need to make sure the programs are all the same in some additional way?

Reply:
The PR is fine as-is. I was just noting a potential issue that could/did happen when unbundling build tools.